lib/elixir/src/elixir_erl_pass.erl

```diff
@@ -361,21 +362,30 @@ translate_map(Meta, Assocs, S) ->
 translate_struct(Meta, Name, {'%{}', _, [{'|', _, [Update, Assocs]}]}, S) ->
   Ann = ?ann(Meta),
   Generated = erl_anno:set_generated(true, Ann),
-  {VarName, _, VS} = elixir_erl_var:build('_', S),
-
-  Var = {var, Ann, VarName},
-  Map = {map, Ann, [{map_field_exact, Ann, {atom, Ann, '__struct__'}, {atom, Ann, Name}}]},
-
-  Match = {match, Ann, Var, Map},
-  Error = {tuple, Ann, [{atom, Ann, badstruct}, {atom, Ann, Name}, Var]},
+  {TUpdate, US} = translate_arg(Update, S, S),
 
-  {TUpdate, US} = translate_arg(Update, VS, VS),
-  {TAssocs, TS} = translate_map(Meta, Assocs, {ok, Var}, US),
+  case elixir_erl:get_type(TUpdate, US) of
+    {struct, Name} ->
+      translate_map(Meta, Assocs, {ok, TUpdate}, US);
+    MaybeMatches when MaybeMatches =:= map orelse MaybeMatches =:= term ->
+      {VarName, _, VS} = elixir_erl_var:build('_', US),
+
+      Var = {var, Ann, VarName},
+      Map = {map, Ann, [{map_field_exact, Ann, {atom, Ann, '__struct__'}, {atom, Ann, Name}}]},
+
+      Match = {match, Ann, Var, Map},
+      Error = {tuple, Ann, [{atom, Ann, badstruct}, {atom, Ann, Name}, Var]},
+
+      {TAssocs, TS} = translate_map(Meta, Assocs, {ok, Var}, VS),
+      {{'case', Generated, TUpdate, [
+        {clause, Ann, [Match], [], [TAssocs]},
+        {clause, Generated, [Var], [], [elixir_erl:remote(Ann, erlang, error, [Error])]}
+      ]}, TS};
+    _WontMatch ->
```

Comment:
This PR is missing something important, which is to consider how `elixir_erl_var:mergev` is going to work for `var_types`. mergev is called to merge scopes before and after calls, such as:

    a = 1
    Foo.bar(a = 2, a)
    a

In the example above, the call will be `Foo.bar(2, 1)`. Luckily, because you are checking the variables after static single assignment, it will always work since the `a`s will have different versions.

The problem though is in code like this:

```elixir
x = 1
case List.flatten([[1, 2, 3]]) do
  [1, 2, 3] ->
    x = 2
  _ ->
    x = 3
end
x
```

which emits this Erlang AST:

```erlang
{block,0,
    [{match,1,{var,1,x@1},{integer,0,1}},
     {'case',2,
         {call,2,
             {remote,2,{atom,0,'Elixir.List'},{atom,2,flatten}},
             [{cons,0,
                  {cons,0,
                      {integer,0,1},
                      {cons,0,{integer,0,2},{cons,0,{integer,0,3},{nil,0}}}},
                  {nil,0}}]},
         [{clause,3,
              [{cons,0,
                   {integer,0,1},
                   {cons,0,{integer,0,2},{cons,0,{integer,0,3},{nil,0}}}}],
              [],
              [{match,4,{var,4,x@2},{integer,0,2}},
               {match,2,{var,0,x@3},{var,0,x@2}},
               {var,4,x@2}]},
          {clause,5,[{var,5,'_'}],[],[{match,6,{var,6,x@3},{integer,0,3}}]}]},
     {var,8,x@3}]}
```

You can see the variable `x@3` is used in two branches. However, your implementation will consider only the assignment from the first branch.

In such cases, we should remove any type information from the assigned type. The logic is around here:

https://github.com/elixir-lang/elixir/blob/master/lib/elixir/src/elixir_erl_clauses.erl#L79-L90

`FinalVars` is the list of variables that will be assigned through multiple branches.

I would recommend removing this clause altogether and let the runtime check fail from the clause above (so we only check for positive cases and optimize them but potential failures are kept as is with runtime checks).